---
title: "CDS changes"
author: "Kent Riemondy RBI"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
source(here::here("R/globals.r"))

fig_dir <- "figs_cds"
out_dir <- "output_files"
dir.create(fig_dir, recursive = TRUE)

```


```{r}
isos <- read_tsv(file.path(out_dir, "all_isoform_annotations.tsv"))
```


## RPFs

Examine RPFs or TE in isoforms. Annotate isoform with riboprofiling evidence if increase post-MZT. 

Show a bar plot wiht # of annotations that are supported by RPFs, Mass Spec, or both.

Pull in ribosome profiling coverage for specific isoform regions enriched post mzt

```{r}
read_bw <- function(bw_fn, 
                    region = NULL,
                    bw_col_types = "cddd",
                    bw_col_names = c("chrom", 
                                     "start",
                                     "end",
                                     "score")){
  
  bw2bg <- Sys.which("/Users/kriemo/bin/kent/bigWigToBedGraph")

  if(bw2bg == "") {
    stop("couldn't find bigWigToBedGraph")
  }
  
  if(is.null(region)){
     res <- system2(bw2bg, 
          c(bw_fn, 
            "stdout"),
          stdout = TRUE) 
  } else {
    
    region[2] <- format(as.numeric(region[2]), 
                         scientific = FALSE)
    region[3] <- format(as.numeric(region[3]), 
                         scientific = FALSE)
        
    res <- system2(bw2bg, 
          c(bw_fn, 
            paste0("-chrom=", region[1]),
            paste0("-start=", region[2]),
            paste0("-end=", region[3]),
            "stdout"),
          stdout = TRUE) 

  } 
  
  if(length(res) == 0){
    res <- tibble()
  } else if (length(res) == 1){
    res <- read_tsv(paste0(res, "\n"), 
                    col_names = bw_col_names, 
                    col_types = bw_col_types)
  } else {
    res <- read_tsv(res, 
                    col_names = bw_col_names, 
                    col_types = bw_col_types)
  }
  res
}
```

```{r}

bw_fns <- dir("../../data/bigwigs/drosophila/EICHHORN/", 
              pattern = "fwd.bw", 
              full.names = TRUE)
```



```{r}
mdata <- read_tsv("~/Projects/mzt-introns/data/raw_data/drosophila/EICHHORN/PRJNA326506.txt")

rpf_mdata <- filter(mdata, str_detect(sample_title,
                                      "_wt_") & str_detect(sample_title, "RPF"))

srr_ids <- basename(bw_fns) %>% 
  str_split(., "_", simplify = TRUE) %>%
  .[, 1]

rpf_bws <- bw_fns[srr_ids %in% rpf_mdata$run_accession]


fwd_bws <- file.path("../../data/bigwigs/drosophila/EICHHORN/",
                 paste0(rpf_mdata$run_accession, "_fwd.bw"))

names(fwd_bws) <- rpf_mdata$sample_title

rev_bws <- file.path("../../data/bigwigs/drosophila/EICHHORN/",
                 paste0(rpf_mdata$run_accession, "_rev.bw"))

names(rev_bws) <- rpf_mdata$sample_title
```


```{r}
iso_annots <- isos

#get intervals

library(kentr)
gtffile <- annots$drosophila$gtf
chromfile <- annots$drosophila$chroms
genomefile <- annots$drosophila$genome

gnome <- read_tsv(chromfile, col_types = "ci---", 
                  col_names= c("chrom", "size"))

gtf <- import(gtffile)
gtf <- as.data.frame(gtf) %>% 
  mutate_if(is.factor, as.character)


iso_ivls <- filter(gtf, 
                  transcript_id %in% iso_annots$featureID,
                  type == "CDS") %>% 
  dplyr::select(chrom = seqnames,
                start,
                end,
                gene_id,
                transcript_id,
                strand) %>% 
  mutate(start = start - 1L)

mat_tx <- filter(iso_annots, tx_class == "maternal")
zyg_tx <- filter(iso_annots, tx_class == "zygotic")

zyg_iso <- filter(iso_ivls,
                  transcript_id %in% zyg_tx$featureID) %>% 
  group_by(gene_id, strand)

mat_iso <- filter(iso_ivls,
                  transcript_id %in% mat_tx$featureID) %>% 
  group_by(gene_id, strand)

zyg_specific_exons <- bed_subtract(zyg_iso, mat_iso) %>% 
  arrange(transcript_id)

mat_specific_exons <- bed_subtract(mat_iso, zyg_iso) %>% 
  arrange(transcript_id)
```


```{r}
fwd_ribo_counts <- map_dfr(fwd_bws, read_bw, .id = "sample") %>% 
  mutate(strand = "+")
rev_ribo_counts <- map_dfr(rev_bws, read_bw, .id = "sample") %>% 
  mutate(strand = "-")

ribo_counts <- bind_rows(fwd_ribo_counts, rev_ribo_counts) %>% 
  bed_sort()

rm(fwd_ribo_counts, rev_ribo_counts)

ribo_counts <- split(ribo_counts, ribo_counts$sample)  
```

```{r}
ribo_summary_zyg <- map_dfr(ribo_counts, 
        ~bed_map(zyg_specific_exons, 
                 .x, 
                 mean_coverage = mean(score, na.rm = TRUE)) %>% 
          mutate(mean_coverage = ifelse(is.na(mean_coverage),
                                        0,
                                        mean_coverage)),
        .id = "source")

ribo_summary_mat <- map_dfr(ribo_counts, 
        ~bed_map(mat_specific_exons, 
                 .x, 
                 mean_coverage = mean(score, na.rm = TRUE)) %>% 
          mutate(mean_coverage = ifelse(is.na(mean_coverage),
                                        0,
                                        mean_coverage)),
        .id = "source")

ribo_summary <- bind_rows(list(
  `Maternal exons` = ribo_summary_mat,
  `Zygotic exons` = ribo_summary_zyg),
  .id = "class")

rename_stages <- c(
  "Stage_11" = "Stage 11",
  "Stage_12" = "Stage 12",
  "Stage_13" = "Stage 13",
  "Stage_14" = "Stage 14",
  "Activated_egg" = "Activated_egg",
  "0-1_h" = "0-1 hr",
  "2-3_h" = "2-3 hr",
  "3-4_h" = "3-4 hr",
  "5-6_h" = "5-6 hr")
  

ribo_summary <- mutate(ribo_summary,
                       source = str_split(source, "_wt", simplify = TRUE) %>% 
                         .[, 1],
                       source_renamed = rename_stages[source],
                       source_renamed = factor(source_renamed, 
                                               levels = rename_stages))

ribo_summary <- ribo_summary %>% 
  group_by(class, gene_id, transcript_id) %>%
  mutate(zscore = (mean_coverage - mean(mean_coverage)) / sd(mean_coverage))
                                       
p <- ggplot(ribo_summary,
       aes(source_renamed, zscore)) +
  geom_boxplot(aes(fill = class), coef = Inf) +
  scale_fill_brewer(palette = "Set1",
                    name = "") +
  labs(x = "",
       y = "Ribosome Profiling\nCoverage (Z-score)") +
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
p
save_plot(file.path(fig_dir, "riboprofiling_coverage.pdf"),
          p,
          base_asp = 1.7)
```

The stage 14 and activated egg RPF data was previoulsy published and reanalyzed by Eichhorn et al. These samples did not have cycloheximide. 
>
Kronja I, Yuan B, Eichhorn SW, Dzeyk K, Krijgsveld J, Bartel DP, Orr-Weaver TL. Widespread changes in the posttranscriptional landscape at the Drosophila oocyte-to-embryo transition. Cell Reports. 2014;7:1495â€“1508. doi: 10.1016/j.celrep.2014.05.002. 
>

```{r}
cap_infinite_vals <- function(x) {
  min_val <- min(x[is.finite(x)], na.rm = TRUE)
  max_val <- max(x[is.finite(x)], na.rm = TRUE)
  ifelse(is.infinite(x) & sign(x) == -1,
         min_val,
         ifelse(is.infinite(x),
                max_val,
                x))
}

# sum up mean exonic coverage 
# drop activated egg as I believe cycloheximide was not used for these samples

ribo_fc_summary <- ribo_summary %>%
  #filter(class == "Zygotic exons") %>% 
  group_by(class, transcript_id, source_renamed) %>% 
  summarize(total_coverage = sum(mean_coverage)) %>% 
  filter(!source_renamed %in% c("Activated_egg",
                                "Stage_14")) %>% 
  mutate(mzt_tpt = ifelse(source_renamed %in% c("3-4 hr","5-6 hr"),
                          "rpfs-post-major-mzt",
                          "rpfs-pre-major-mzt")) %>% 
  group_by(class, transcript_id, mzt_tpt) %>% 
  summarize(mean_tpt_coverage =  mean(total_coverage)) %>% 
  pivot_wider(names_from = mzt_tpt,
              values_from = mean_tpt_coverage) %>% 
  mutate(rpf_logfc = log2(`rpfs-post-major-mzt`) - log2(`rpfs-pre-major-mzt`))

ribo_fc_summary$rpf_logfc <- cap_infinite_vals(ribo_fc_summary$rpf_logfc)

ribo_fc_summary <- ribo_fc_summary %>%
  ungroup() %>% 
  mutate(ribo_seq_change = ifelse(rpf_logfc > 1, 
                                  "increased RPFs",
                                  ifelse(rpf_logfc < 1,
                                         "decreased RPFs",
                                         "similar RPFs"))) %>% 
  select(-class)
isos <- left_join(isos,
                  ribo_fc_summary, 
                  by = c("featureID" = "transcript_id"))
```

## Mass spec

```{r}
mass_spec_evidence <- filter(isos, 
                             !same_CDS, 
                             !is.na(seq),
                             tx_class == "zygotic") %>% 
  select(pep_tx, featureID) %>% 
  unique() %>% 
  mutate(peptide_evidence = !is.na(pep_tx))

p <- ggplot(mass_spec_evidence, aes("1")) +
  geom_bar(aes(fill = peptide_evidence))  +
  scale_fill_viridis_d(name = "Peptides Observed \n in Mass Spec") +
  labs(x = "",
       y = "# of proteins") + 
  theme(axis.text.x = element_blank())

p

save_plot(file.path(fig_dir, "mass_spec_overlap.pdf"),
          p,
          base_asp = 1)

```

## Protein domain analysis

Examine zygotic isoforms and count # of isoforms with domain gain, loss, or both.

```{r}

if(!file.exists("../../docs/pfam_ens_biomart.tsv")){
  library(biomaRt)
  ens <- useMart("ensembl", dataset="dmelanogaster_gene_ensembl")
  
 # iso_tx_idx <- isos$featureID %>% unique() 
  iso_prot <- getBM(attributes = c("flybase_transcript_id","pfam","pfam_start", "pfam_end"), 
                    filters = "flybase_transcript_id",
                    values = unique(gtf$transcript_id), 
                    mart = ens)
  
  iso_prot[iso_prot == ""] <- NA
  
  write_tsv(iso_prot, "../../docs/pfam_ens_biomart.tsv")
} 

iso_prot <- read_tsv("../../docs/pfam_ens_biomart.tsv")

if(!file.exists("../../docs/pfam_id_map.txt")){
  download.file("ftp://ftp.ebi.ac.uk/pub/databases/Pfam/mappings/pdb_pfam_mapping.txt", 
              "../../docs/pfam_id_map.txt")
}

pfam_map <- read_tsv("../../docs/pfam_id_map.txt") %>% 
  mutate(pfam = str_remove(PFAM_ACC, "\\.[0-9]+$")) 

# 
iso_prot <- left_join(iso_prot, pfam_map, by = "pfam")
iso_prot <- dplyr::select(iso_prot, flybase_transcript_id:pfam_end, PFAM_desc)
iso_prot <- dplyr::group_by(iso_prot, 
                           flybase_transcript_id, 
                           pfam, 
                           pfam_start, 
                           pfam_end) %>% 
  dplyr::summarize(PFAM_desc = dplyr::first(PFAM_desc))


iso_prot_summary <- left_join(isos, 
          iso_prot,
          by = c("featureID" = "flybase_transcript_id")) %>% 
  mutate(pfam_aa = str_sub(aa, pfam_start, pfam_end))

```


```{r}
pfam_iso_domains <- filter(iso_prot_summary, !same_CDS) %>% 
                           dplyr::select(tx_class, 
                                         gene_name,
                                         featureID, 
                                         pfam,
                                         pfam_aa)

# see: https://stackoverflow.com/a/55697861
setdiff_duplicates <- function(x, y) {
  # get freq tables for x and y
  x.tab <- table(x)
  y.tab <- table(y)
  # if a value is missing in y then set its freq to zero
  y.tab[setdiff(names(x.tab), names(y.tab))] = 0
  y.tab <- y.tab[names(y.tab) %in% names(x.tab)]
  # get the difference of x and y freq and keep if > 0
  diff.tab <- x.tab[order(names(x.tab))] - y.tab[order(names(y.tab))]
  diff.tab <- diff.tab[diff.tab > 0]
  # output vector of x values missing in y
  unlist(
    lapply(names(diff.tab), function(val) {
      rep(val, diff.tab[val])
    }), 
    use.names = F)
}

set_relationships <- function(x, y){
  
  x_missing <- is.na(x)
  y_missing <- is.na(y)
  
  if(any(x_missing) && length(x_missing) > 1) {
      stop("x has NA and other values not supposed to happen")
  }

  if(any(y_missing) && length(y_missing) > 1) {
      stop("x has NA and other values not supposed to happen")
  }

  x_missing <- any(x_missing)
  y_missing <- any(y_missing)
  if(x_missing & y_missing){
    return("same")
  } else if (x_missing){
    return("gain")
  } else if (y_missing) {
    return("loss")
  }
  
  # check values first
  element_loss <- length(setdiff_duplicates(x, y)) > 0
  element_gain <- length(setdiff_duplicates(y, x)) > 0  
  mixed <- element_loss & element_gain
  
  if(mixed){
    return("gain_and_loss")
  } else if (element_loss){
    return("loss")
  } else if (element_gain){
    return("gain")
  } else if (identical(sort(x), sort(y))){
    return("same")
  } else {
    stop("not sure this should happen")
  }
}


find_domain_similarities <- function(tx_list){
  
  map_dfr(tx_list, function(tx){
    txs <- split(tx, tx$tx_class)
    
    domains_present <- set_relationships(txs$maternal$pfam, txs$zygotic$pfam)
    to_check <- intersect(txs$maternal$pfam, txs$zygotic$pfam) 

    ma <- txs$maternal$pfam_aa
    names(ma) <- txs$maternal$pfam

    za <- txs$zygotic$pfam_aa
    names(za) <- txs$zygotic$pfam
  
    domain_changes <- c(!za %in% ma, !ma%in% za)
    
    tx[["domain_summary"]] <- domains_present
    tx[["domain_seq_changed"]] <- domain_changes
    tx
  })
}

pfam_iso_domains <- pfam_iso_domains %>% 
  group_by(gene_name) %>%
  arrange(desc(tx_class), 
          .by_group = TRUE)


tx_list <- split(pfam_iso_domains, pfam_iso_domains$gene_name) 

domain_summary_out <- find_domain_similarities(tx_list)

pretty_rename <- c(
  "domain sequence changed" = "Domain sequence changed",
  "same" = "Other CDS change",
  "loss" = "Domain loss",
  "gain" = "Domain gain")

plt_dat <- select(domain_summary_out, 
       gene_name,
       domain_summary,
       domain_seq_changed) %>% 
  group_by(gene_name) %>% 
  mutate(domain_seq_changed = any(domain_seq_changed),
         domain_seq_changed = ifelse(domain_summary == "same",
                                     domain_seq_changed,
                                     NA)) %>% 
  unique() %>% 
  ungroup() %>% 
  mutate(domain_summary = ifelse(domain_summary == "same" & domain_seq_changed,
                                 "domain sequence changed",
                                 domain_summary),
         domain_summary = pretty_rename[domain_summary])

p <- ggplot(plt_dat,
       aes("1")) +
    geom_bar(aes(fill = domain_summary)) +
  scale_fill_viridis_d(name = "") +
  labs(x = "",
       y = "# of proteins") + 
  theme(axis.text.x = element_blank())
p
save_plot(file.path(fig_dir, "cds_domain_change_summary.pdf"), p,
          base_asp = 1.2)
```

Make some summary plots

```{r}
  
euler_input <- isos %>% 
  filter(!is.na(seq)) %>% #drop transcripts with no CDS
  select(gene_name, 
         same_CDS,
                 same_five_prime_utr,
                 same_three_prime_utr) %>% 
  unique() %>% 
  column_to_rownames("gene_name") 
euler_input <- !euler_input
colnames(euler_input) <- c("CDS altered",
                           "5'UTR altered",
                           "3'UTR altered")
fit <- eulerr::euler(euler_input)

plot(fit,
     quantities = list(fontsize = 16),
     labels = list(fontsize = 20),
     fill = list(fill = viridisLite::viridis(3), 
                 alpha = 0.50))

pdf(file.path(fig_dir, "utr_cds_overlap_isoforms.pdf"))
plot(fit,
     quantities = list(fontsize = 16),
     labels = list(fontsize = 20),
     fill = list(fill = viridisLite::viridis(3), 
                 alpha = 0.50))
dev.off()
```

## Show some examples of domain changes and coverage  



## uORFs 

Is there any evidence of uORF activity in genes with changed 5'UTRs?

```{r}
gtf_tbl <- tidy_gtf(annots$drosophila$gtf) 

utr5_annot <- gtf_tbl %>% 
  filter(transcript_id %in% isos$featureID,
         type == "five_prime_utr") %>% 
  group_by(strand)


mdata <- read_tsv("../2019-11-21_rpfs/updated_eichhorn_mdata.txt") %>% 
  mutate(stage = str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[1:2], collapse = "_")),
         gt = str_split(sample_title, "_") %>% 
           map_chr(~.x[3]),
         treatment = str_split(sample_title, "_") %>% 
           map_chr(~.x[4]),
         treatment = ifelse(treatment =="No",
                            str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[4:5], collapse = "_")),
           treatment),
         type = str_match(sample_title, 
                          "RPF|RNA|PAL")[, 1]) 


mdata <- mdata %>% 
  filter(type != "PAL",
         gt == "wt")

mdata <- mdata %>% 
  arrange(sample_title, 
          type)

ribo_mdata <- filter(mdata, type == "RNA")

fc_files  <- str_c("../../data/featurecounts/drosophila/EICHHORN/",
                   ribo_mdata$run_accession, 
                   "_5utr_tx_counts.tsv")
fc_5utr_mat <- map(fc_files,
    ~read_tsv(.x, comment = "#") %>% 
      select(Geneid, 7)) %>% 
  Reduce(function(x,y){inner_join(x,
                                  y, 
                                  by = "Geneid")},
         x = .)

colnames(fc_5utr_mat) <- basename(colnames(fc_5utr_mat)) %>% 
  str_remove("_sorted.bam")

# read in summary files and sum up assigned,_NoFeature, and noambiguity for
# total read count for normalization.

fc_summary_files  <- str_c("../../data/featurecounts/drosophila/EICHHORN/",
                   ribo_mdata$run_accession, 
                   "_5utr_tx_counts.tsv.summary")
names(fc_summary_files) <- ribo_mdata$run_accession
fc_5utr_summary <- imap(fc_summary_files,
                       ~read_tsv(.x) %>% 
                         .[c(1, 10, 12), 2, drop = TRUE] %>% 
                         sum())
                         
tmp_mat <- fc_5utr_mat[2:length(fc_5utr_mat)]

for(i in colnames(tmp_mat)){
  tmp_mat[, i] <- 1e6 * (tmp_mat[, i] / fc_5utr_summary[[i]])
}

fc_5utr_mat[2:length(fc_5utr_mat)] <- tmp_mat

iso_5utr_rpfs <- filter(fc_5utr_mat, Geneid %in% isos$featureID)

colnames(iso_5utr_rpfs) <-  left_join(
  tibble(run_accession = colnames(iso_5utr_rpfs)), 
  ribo_mdata) %>% 
  mutate(sample_title = ifelse(is.na(sample_title),
                               "Geneid",
                               sample_title)) %>% 
  pull(sample_title)

utr_counts <- isos %>% 
  dplyr::select(featureID, tx_class, gene_name, starts_with("same")) %>% 
  left_join(iso_5utr_rpfs, by = c("featureID" = "Geneid")) 

write_tsv(utr_counts, "transcript_level_5utr_counts.tsv.gz")
```

## TE

Is there any evidence for TE changes? 



```{r}
gtf <- tidy_gtf(annots$drosophila$gtf)

cds_gtf <- filter(gtf, type == "CDS")

cds <- txcompare::gtf_to_seq(cds_gtf, annots$drosophila$genome, type_field = "CDS")

cds <- mutate(cds, seq = str_sub(seq, start = 51))
cds <- filter(cds, nchar(seq) > 50)
write_fasta(cds, "ribosome_profiling_cds.fa", header_col = "transcript_id")

mdata <- read_tsv("../2019-11-21_rpfs/updated_eichhorn_mdata.txt") %>% 
  mutate(stage = str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[1:2], collapse = "_")),
         gt = str_split(sample_title, "_") %>% 
           map_chr(~.x[3]),
         treatment = str_split(sample_title, "_") %>% 
           map_chr(~.x[4]),
         treatment = ifelse(treatment =="No",
                            str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[4:5], collapse = "_")),
           treatment),
         type = str_match(sample_title, 
                          "RPF|RNA|PAL")[, 1]) 

mdata <- mdata %>% 
  filter(type != "PAL",
         gt == "wt")

mdata <- mdata %>% 
  arrange(sample_title, 
          type)
```


```{r}
mdata <- read_tsv("../2019-11-21_rpfs/updated_eichhorn_mdata.txt") %>% 
  mutate(stage = str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[1:2], collapse = "_")),
         gt = str_split(sample_title, "_") %>% 
           map_chr(~.x[3]),
         treatment = str_split(sample_title, "_") %>% 
           map_chr(~.x[4]),
         treatment = ifelse(treatment =="No",
                            str_split(sample_title, "_") %>% 
           map_chr(~str_c(.x[4:5], collapse = "_")),
           treatment),
         type = str_match(sample_title, 
                          "RPF|RNA|PAL")[, 1]) 

mdata <- mdata %>% 
  filter(type != "PAL",
         gt == "wt")

mdata <- mdata %>% 
  arrange(sample_title, 
          type)
```


```{r}
rna_files <- file.path(data_dir,
                            "salmon", 
                            "drosophila",
                            "EICHHORN",
                            "ribosome_profiling_cds",
                            filter(mdata, type == "RNA") %>%
                         pull(run_accession),
                            "quant.sf")

rpf_files <- file.path(data_dir,
                            "salmon", 
                            "drosophila",
                            "EICHHORN",
                            "ribosome_profiling_cds",
                            filter(mdata, type == "RPF") %>%
                         pull(run_accession),
                            "quant.sf")

txirna <- tximport(rna_files,
                type="salmon", 
                txOut=TRUE,
                countsFromAbundance="scaledTPM")

txirpf <- tximport(rpf_files,
                type="salmon", 
                txOut=TRUE,
                countsFromAbundance="scaledTPM")

all(rownames(txirna$abundance) == rownames(txirpf$abundance))


te <- txirpf$abundance / txirna$abundance

out_col_names <- filter(mdata, type == "RNA") %>%
                         pull(stage)

colnames(te) <- out_col_names 

te <- tibble::rownames_to_column(as.data.frame(te), "transcript_id")

te <- dplyr::select(
  te,
  transcript_id,
  Stage_11:Stage_14,
  Activated_egg,
  `0-1_h`:`5-6_h`
)

rpfs <- txirpf$abundance 
colnames(rpfs) <- filter(mdata, type == "RNA") %>%
                         pull(stage)

rpfs <- rpfs %>% 
  as.data.frame() %>% 
  rownames_to_column("transcript_id") %>% 
  dplyr::select(
  transcript_id,
  Stage_11:Stage_14,
  Activated_egg,
  `0-1_h`:`5-6_h`
)
```


Get map of tx_ids not in tables due to having the same CDS
```{r}
dup_txs <- read_tsv("../../dbases/drosophila/salmon/ribosome_profiling_cds/duplicate_clusters.tsv")
```


Add additional rows for duplicate tx's

```{r}
dup_tes <- left_join(dup_txs, te, by = c("RetainedTxp" = "transcript_id")) %>% 
  dplyr::select(transcript_id = DuplicateTxp,
         representative_cds_transcript = RetainedTxp, 
         everything())

te <- mutate(te, 
             representative_cds_transcript = transcript_id)

te <- bind_rows(dup_tes, te)

write_tsv(te, "transcript_level_te.tsv.gz")


dup_rpfs <- left_join(dup_txs, rpfs, 
                      by = c("RetainedTxp" = "transcript_id")) %>% 
  dplyr::select(transcript_id = DuplicateTxp,
         representative_cds_transcript = RetainedTxp, 
         everything())

rpfs <- mutate(rpfs, 
             representative_cds_transcript = transcript_id)

rpfs <- bind_rows(dup_rpfs, rpfs)

write_tsv(rpfs, "transcript_level_rpfs.tsv.gz")

```

```{r}
#calc tpm fold changes and examine across dev time 




```





```{r}
iso_annots <- read_tsv("../2019-06-09_isoform/isoform_summaries.tsv.gz")

iso_utr_changes <- iso_annots %>% 
  filter(!same_CDS) 

te_iso_changes <- iso_annots %>%
  select(tx_class, featureID, gene_name) %>% 
  left_join(te, by = c("featureID" = "transcript_id")) 

#te_changes[is.na(te_changes)] <- 0
te_iso_changes <- na.omit(te_iso_changes)
sample_order <- dplyr::select(te, Stage_11:`5-6_h`) %>% 
  colnames()
  
te_changes <- te_iso_changes %>% 
  pivot_longer(cols = -(tx_class:representative_cds_transcript)) %>% 
  mutate(sample = factor(name, levels = sample_order),
         value = log2(value))



ggplot(te_changes, aes(sample, value)) + 
  geom_violin(aes(fill = tx_class), 
              position = position_dodge()) +
  labs(x = "",
       y = "Translation Efficiency (log2)") +
  scale_fill_brewer(palette = "Set1", 
                    name = "") +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
```


```{r}

rpf_fc_files <- file.path(data_dir,
                            "featurecounts", 
                            "drosophila",
                            "EICHHORN",
                            filter(mdata, type == "RPF") %>%
                         pull(run_accession) %>% 
                           str_c(., "_5utr_tx_counts.tsv"))

rpf_5utr_counts <- map(rpf_fc_files, 
                       read_tsv, 
                       skip = 1) %>% 
  Reduce(function(x, y) {
    inner_join(x,y, 
               by = c("Geneid",
                      "Chr",
                      "Start",
                      "End",
                      "Strand",
                      "Length"))}, 
    .)

libs <- colnames(rpf_5utr_counts) %>% 
  basename() %>% 
  str_remove("_sorted.bam") %>% 
  str_subset("SRR")

ids <- inner_join(tibble(run_accession = libs),
           mdata,
           by = "run_accession") %>% 
  pull(stage)


colnames(rpf_5utr_counts) <- c(c("Geneid",
                      "Chr",
                      "Start",
                      "End",
                      "Strand",
                      "Length"),
                   ids)

rpf_5utr_counts <- rpf_5utr_counts %>% 
  dplyr::select(transcript_id = Geneid, 
                Chr:Length,
                Stage_11:Stage_14,
  Activated_egg,
  `0-1_h`:`5-6_h`)


write_tsv(rpf_5utr_counts, "5putr_rpf_counts.tsv.gz")
```

```{R}
a <- read_tsv("transcript_level_5utr_counts.tsv.gz")
```



## Plot transcripts with riboprofiling and RNA-seq coverage

## Representative transcripts

Show coverage profiles from transcripts that swtich
```{r, eval =FALSE}
base_url <- "http://amc-sandbox.ucdenver.edu/User33/rissland/bws/"
sras <- filter(mdata, strain == "w1118") %>% pull(Run)
ids <- filter(mdata, strain == "w1118") %>% pull(developmental_stage)

bws <- c("_fwd.bw", "_rev.bw")
names(bws) <- c("+", "-")

bw_fns <- map(bws, function(x) {
  out <- paste0(base_url, sras, x)
  names(out) <- ids
  out
}) 


gtf_obj <- import(annots$drosophila$gtf)

tx_summary_plots <- function(bws, 
                             gene,
                             isoform_annotation_df,
                             ...){
  
  to_plot <- filter(additional_annots, gene_name == gene)
  strand <- unique(to_plot$strand)[1]

  features <- unique(to_plot$featureID)
  plot_tx_coverage(bw_fns[[strand]], 
                 features,
                 gtf_obj,
                 scale_y = FALSE,
                 cex.main = 0.75,
                 rotation.title=0,
                 ...
                 )
}

to_plot <- c("l(2)k09913",
  "PNUTS",
  "cnn",
  "Zn72D",
  "Rb97D",
  "pyd",
  "CG8331",
  "Pk92B",
  "yata",
  "hts",
  "bl",
  "eIF-2gamma",
  "Npc1a",
  "Su(var)2-10",
  "toc",
  "Hph",
  "blot",
  "Mhcl",
  "SC35",
  "CG4658",
  "arm",
  "Nmnat",
  "Adar",
  "Plip",
  "lola")

tx_figs <- file.path(fig_dir, "tx_figs")
dir.create(tx_figs, showWarnings = FALSE)

walk(to_plot,
    ~{
      pdf(file.path(tx_figs,
                    str_c(.x, "_tx_coverage.pdf")),
          width = 6,
          height = 3)
      tx_summary_plots(bw_fns, gene = .x, 
                 isoform_annotation_df = additional_annots,
                 title.width = 1.5,
                 sizes = c(0.1, rep(0.25, 5)),
                 fontsize = 8,
                 cex.title = 1)
      dev.off()
    })
```
