---
title: "Intron metagenes"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message = FALSE, warning = FALSE, include = FALSE}
source(here::here("R/globals.r"))
library(doParallel)
fig_dir <- "figs"
dir.create(fig_dir, recursive = TRUE)
```


## Examine read distribution surrounding intron-exon junctions
  
### generate intron positions drosophila

```{r intron_pos}

generate_splice_regions <- function(genome_file, gtf_file){
  genome <- read_tsv(genome_file, 
                     col_types = c("ci---"), 
                     col_names = c("chrom", "size"))
  
   bed <- tidy_gtf(gtf_file) %>% 
     filter(type == "exon") %>% 
     dplyr::select(chrom, start, end, transcript_id, gene_id, score, strand)
  
  # generate introns
  intron_bed <- bed %>% 
    group_by(transcript_id, gene_id) %>% 
    arrange(start, .by_group = T) %>% 
    mutate(.start = end, 
           .end = lead(start),
          start = .start, 
          end = .end) %>% 
     dplyr::select(-.start, -.end) %>% 
     dplyr::filter(!is.na(start),
           !is.na(end),
           start < end)
  
  # add intron number as score by strand
  intron_bed <- intron_bed %>% 
    mutate(score = ifelse(strand == "-",
                          rev(row_number()), 
                          row_number())) %>% 
    ungroup()
  
  
  ss_3p <- intron_bed %>% 
    group_by(transcript_id) %>%
    arrange(score, .by_group = TRUE) %>% 
   # dplyr::slice(1:(n() - 1)) %>% 
    ungroup() %>% 
    mutate(start = ifelse(strand == "-", 
                          start,
                          end - 1),
           end = ifelse(strand == "-",
                        start + 1, 
                        end))
  
  ss_5p <- intron_bed %>% 
    group_by(transcript_id) %>%
    arrange(score, .by_group = TRUE) %>% 
   # dplyr::slice(2:n()) %>% 
    ungroup() %>% 
    mutate(start = ifelse(strand == "-", 
                          end - 1,
                          start),
           end = ifelse(strand == "-",
                        end, 
                        start + 1))
  
  splice_pos <- list(ss_5p, ss_3p)
  splice_pos <- map(splice_pos, 
                    ~dplyr::select(.x, -transcript_id) %>% 
                      unique())
  
  # make intervals +/- 50bp 
  splice_slop <- map(splice_pos, 
                       ~bed_slop(.x, genome, both = 50))
    
  # single nucleotide coverage
  splice_windows_plus <- map(splice_slop,
                          ~dplyr::filter(.x, strand == "+") %>%  
                            bed_makewindows(win_size = 1))
   
  splice_windows_minus <- map(splice_slop,
                          ~dplyr::filter(.x, strand == "-") %>%  
                            bed_makewindows(win_size = 1,
                                           reverse = TRUE))
  
  splice_windows <- map2(splice_windows_plus, 
                         splice_windows_minus,
                         ~bind_rows(.x, .y))
  splice_windows
}

splice_regions <- generate_splice_regions(annots$drosophila$chroms,
                                          annots$drosophila$gtf)

gtf_dat <- tidy_gtf(annots$drosophila$gtf)

```

## Get public gene sets 

```{r}
# derenzis genes
renzis_zyg_genes <- read_delim(file.path(project_dir, "docs",
                                   "renzis_all_zygotic.txt"),
                           delim = " ", 
                           col_names = FALSE)

renzis_early_genes <- read_delim(file.path(project_dir, "docs",
                                   "renzis_early_zygotic.txt"),
                           delim = " ", col_names = FALSE, skip = 1)

renzis_mzt_genes <- read_delim(file.path(project_dir, "docs",
                                   "renzis_mzt.txt"),
                           delim = " ", col_names = FALSE, skip = 1)


# eisen genes

if(!file.exists(file.path(project_dir, 
                        "docs", 
                        "eisen_data.tsv"))) {
  download.file("https://ndownloader.figshare.com/files/400050",
                file.path(project_dir, 
                          "docs", 
                          "eisen_data.tsv"))
}

edata <- read_tsv(file.path(project_dir, 
                        "docs", 
                        "eisen_data.tsv"))

genename2id <- filter(gtf_dat, type == "gene") %>% 
  dplyr::select(gene_name, gene_id) %>% 
  unique()
edata_genes <- dplyr::select(edata, NAME, CLASS) %>% 
  left_join(genename2id, by = c("NAME" = "gene_name")) %>% 
  na.omit() %>% 
  split(., .$CLASS)

bartel_zyg <- expr_genes <- read_excel("../../docs/bartel_2019_s1.xlsx", skip = 2) %>% 
  left_join(gtf_dat, by = c("GENES" = "gene_name")) %>% 
  pull(gene_id) %>% 
  unique()

public_gene_sets <- list(
  derenzis_early_zygotic = renzis_early_genes$X2,
  derenzis_pure_zygotic = renzis_zyg_genes$X2,
  derenzis_mat_zygotic = renzis_mzt_genes$X2,
  derenzis_any_zygotic = union(renzis_zyg_genes$X2,
                                   renzis_mzt_genes$X2),
  eisen_pure_zygotic = edata_genes$zyg$gene_id,
  eisen_mat_zygotic = edata_genes$matzyg$gene_id,
  eisen_any_zygotic = union(edata_genes$zyg$gene_id, 
                                edata_genes$matzyg$gene_id),
  eisen_pure_mat = edata_genes$mat$gene_id,
  bartel_zyg = bartel_zyg
) %>% 
  map(na.omit)
```

```{r}
map(public_gene_sets, ~length(unique(.x)))
```

## generate coverage summaries

```{r mdata}
mdata1 <- read_tsv(file.path(data_dir, 
                             "raw_data", 
                             "drosophila",
                             "EICHHORN",
                             "GSE83616_run_info_geo.txt"))
mdata2 <- read_tsv(file.path(data_dir, 
                             "raw_data", 
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))


#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("eichhorn", "rissland")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```


```{r drosophila_windows}
compute_coverage <- function(bw, splice_windows_list){

  splice_windows <- map(splice_windows_list, ~group_by(.x, strand))
  bw <- group_by(bw, strand)
  
  splice_coverage <- map(splice_windows,
                         ~bed_map(.x,
                                  bw, 
                                  total_coverage = mean(score, na.rm = T)))
  
  splice_summary <- map(splice_coverage, 
                        ~group_by(.x, .win_id) %>% 
                          summarize(total_signal = sum(total_coverage, 
                                                       na.rm = T)))
  
  res <- bind_rows(splice_summary, .id = "type")
  res 
}

bws <- dir(file.path(data_dir,
                     "bigwigs",
                     "drosophila",
                     "star"), 
           full.names = TRUE,
           pattern = ".bw",
           recursive = T)
sras <- str_match(bws, "SRR[0-9]+")[, 1] %>% 
  unique() 
sras <- sras[sras %in% mdata$Run]
bws <- bws[str_match(bws, "SRR[0-9]+")[, 1] %in% sras]

fwd_bws <- str_subset(bws, "fwd.bw")
rev_bws <- str_subset(bws, "rev.bw")

bw_info <- zip(fwd_bws, 
               rev_bws,
               "+",
               "-")

intron_metagene <- function(list_of_bws, splice_windows) {
  #list_of_bws <- list(bw_pos, bw_neg)
  #splice_windows <- list of bed_intervals for 5p and 3p splice site
  bw_fwd <- read_bigwig(list_of_bws[[1]], list_of_bws[[3]])
  bw_rev <- read_bigwig(list_of_bws[[2]], list_of_bws[[4]])
  
  bw_combined <- bind_rows(bw_fwd, bw_rev)
  res <- compute_coverage(bw_combined, splice_windows)
  res
}

# for some reason purrr::map would freeze if using foreach in map call. 
all_res <- list()
for (i in seq_along(public_gene_sets)){
  message(names(public_gene_sets)[i])
  splice_windows <- map(splice_regions, 
                                            ~filter(.x, gene_id %in% public_gene_sets[[i]]))
                      n <- 3
                      registerDoParallel(n)
                      all_res[[i]] <- foreach(i = bw_info,
                                     .packages = c("valr", "tidyverse", "rtracklayer"),
                                     .export = c("intron_metagene", 
                                                 "read_bigwig",
                                                 "splice_windows"))  %dopar% {
                                                   intron_metagene(i, splice_windows)
                                                 }
}
names(all_res) <- names(public_gene_sets)

```


## Plot coverage
```{r plot}
dev_stages_ordered <- c(
  "Stage 11 oocyte",
  "Stage 12 oocyte",
  "Stage 13 oocyte",
  "Stage 14 oocyte",
  "Activated egg",
  "0-1 hr embryo",
  "1-2 hr embryo",
  "2-3 hr embryo",
  "3-4 hr embryo",
  "4-5 hr embryo",
  "5-6 hr embryo")

sras <- basename(names(bw_info)) %>% str_remove("_fwd.bw") 

# scale everything to 1
all_plts <- map(all_res, function(res){
  plot_ids <- left_join(tibble(Run = sras), 
                        mdata,
                        by = "Run") %>% 
    mutate(id = paste(study, developmental_stage, strain, sep = "_")) %>% 
    pull(id)
  
  names(res) <- plot_ids
  res <- bind_rows(res, .id = "id")
  res <- res %>% 
    tidyr::separate(id, c("study", "stage", "genotype"), sep = "_") %>% 
    group_by(study, stage, type, .win_id) %>% 
    summarize(total_signal = sum(total_signal)) %>% 
    group_by(study, type, stage) %>% 
    mutate(prop_of_max = total_signal / max(total_signal)) %>% 
    ungroup() %>% 
    mutate(stage = factor(stage, levels = dev_stages_ordered))
  
  
  new_facet_names <- c(
    `1` = "5p-Splice-Site",
    `2` = "3p-Splice-Site"
  )
  
  x_labels <- seq(-50, 50, by = 25)
  x_breaks <- seq(1, 101, by = 25)
  
  plts <- map(unique(res$study),
              ~ggplot(dplyr::filter(res, study == .x),
                      aes(.win_id, prop_of_max)) + 
                geom_line(aes(colour = stage), size = 0.75, alpha = 0.75) + 
                facet_grid(~type, labeller = as_labeller(new_facet_names)) +
                scale_colour_viridis(discrete = TRUE,
                                     name = "") + 
                scale_x_continuous(labels = x_labels, breaks = x_breaks) +
                labs(y = "Proportion of maximum signal",
                     x = "Nucleotides from the splice site"))
  
  names(plts) <- unique(res$study)
  plts
})

dir.create(file.path(fig_dir, "intron_metagenes"))
imap(all_plts,
     function(plts, study){
       imap(plts, function(x, y){
       save_plot(file.path(fig_dir, "intron_metagenes", 
                                str_c("genes_from_", study,"_", y, ".pdf")), 
                      x, base_width = 6.5)})
       })
all_plts
saveRDS(all_res, file.path(fig_dir, "intron_metagenes","drosophila_plotting_dat.rds"))
```



```{r}
sessioninfo::session_info()
```