''' rules to count reads overlapping genes '''
from pathlib import Path

def _strand_settings(wildcards):
  try:
    strand = CMD_CONFIG["stranded"]
  except:
    sys.exit(f"stranded settings for {id} not found ")

  if strand == "forward":
      setting = 1
  elif strand == "reverse":
      setting = 2
  else:
      setting = 0

  return setting


rule fc_introns_exons:
  input:
    bam = "{data}/star/{sample}/{sample}_sorted.bam",
  output:
    "{data}/featurecounts/{sample}_intron_exon_counts.tsv",
  params:
    in_annots = INTRONS,
    ex_annots = EXONS,
    strand = _strand_settings,
    r = os.path.join(RSRC, "fcount_introns_exons.R"),
    out_pre = "{data}/featurecounts/{sample}"
  log:
    "{data}/logs/featurecounts/{sample}_intron_summary.txt"
  threads:
    12
  resources: 
      all_threads=12,
      mem_mb=MEM_ALLOC["DEFAULT"]
  shell:
   """
   Rscript --vanilla \
           {params.r} \
           {input} \
           {params.in_annots} \
           {params.ex_annots} \
           {params.strand} \
           {output} \
           {threads}
   """ 
