---
title: "Xenopus T.  MZT analysis"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message = FALSE, warning = FALSE, include = FALSE}
library(here)
source(here('results', '2022-06-21-revisions', 'src',  'globals.r'))
fig_dir <- "figs"
species <- "xenopus_t"
dir.create(fig_dir, showWarnings = FALSE)
```


## Approach


 
## get metadata

```{r get_metadata}
mdata <- read_tsv(file.path(data_dir, "raw_data", "xenopus_t", 
                            "OWENS", "PRJNA275011.txt"))
 
mdata <- separate(mdata,
                sample_title,
                c("clutch",
                  "lib_type",
                  "timepoint",
                  "timetype"),
                sep = "_",
                remove = F)

rzero_samples <- dir(file.path(data_dir,
                   "owens_ribo",
                   "salmon_bt2_masked", 
                   "eisa_masked"))

rzero_mdata <- filter(mdata, run_accession %in% rzero_samples)
rzero_mdata
```


```{r}
polya_samples <- dir(file.path(data_dir,
                   "owens_polya",
                   "salmon_bt2_masked", 
                   "eisa_masked"))

polya_mdata <- filter(mdata, run_accession %in% polya_samples)
polya_mdata

mdata <- bind_rows(rzero_mdata, polya_mdata)
```

```{r }
gtf <- tidy_gtf(annots[[species]]$gtf)

gene2tx <- gtf %>%
   filter(type == "exon") %>%
   select(gene_id, transcript_id) %>%
   na.omit() %>%
   unique()

gene2intron <- read_tsv(here("dbases", "owens_ribo", "eisa", "tx2gene.tsv"))
```


## Examine pre-mRNA ratios

```{r read_in_salmon}
fa_types <- c(
  "eisa_masked"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "owens_ribo",
                        "salmon_bt2_masked", 
                        .x,
                        rzero_mdata$run_accession,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- rzero_mdata$run_accession
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "run_accession"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 rzero_mdata,
                by = c("run_accession"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "-I[0-9]*$"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

dat <- mutate(dat,
              tmp_id = ifelse(count_type == "intron",
                               str_remove(Name, "-I[0-9]*$"),
                               Name))
# add in geneid attribute
tpm_dat_rzero <- left_join(dat, gene2tx, 
                 by = c("tmp_id" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

```

```{r read_in_salmon_polya}
files <- map(fa_types, 
             ~file.path(data_dir,
                        "owens_polya",
                        "salmon_bt2_masked", 
                        .x,
                        polya_mdata$run_accession,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- polya_mdata$run_accession
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "run_accession"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 polya_mdata,
                by = c("run_accession"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "-I[0-9]*$"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

dat <- mutate(dat,
              tmp_id = ifelse(count_type == "intron",
                               str_remove(Name, "-I[0-9]*$"),
                               Name))
# add in geneid attribute
tpm_dat_polya <- left_join(dat, gene2tx, 
                 by = c("tmp_id" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

```

```{r}
tpm_dat <- bind_rows(tpm_dat_rzero, tpm_dat_polya)
# sum primary and mature TPMs for each gene 
summary_dat <-  dplyr::select(tpm_dat, TPM, gene_id, 
              sample_title, count_type, lib_type, timepoint) %>% 
  group_by(gene_id, sample_title, count_type, timepoint, lib_type) %>% 
  dplyr::summarize(total_tpm = sum(TPM, na.rm = F)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- summary_dat %>% 
  spread(count_type, total_tpm) %>% 
  mutate(primary_ratio  = intron / (intron + exon))

## calc mean primary ratio
tpm_means <- tpm_summary %>% 
  group_by(gene_id, timepoint, lib_type) %>% 
  summarize(mean_primary = mean(primary_ratio, na.rm = T)) 

tpm_means <- mutate(tpm_means, timepoint = factor(timepoint))

tpm_means <- dplyr::ungroup(tpm_means)



stage_col_palette <- c("#00A875", "#0072BC")
stage_cols <- stage_col_palette[1 + (levels(tpm_means$timepoint) > "4.5")]
names(stage_cols) <- levels(tpm_means$timepoint)

p <- ggplot(tpm_means, aes(timepoint, mean_primary)) +
  geom_boxplot(aes(fill = timepoint), outlier.shape = NA, coef = 1e10) + 
  facet_wrap(~lib_type) +
  labs(x = "",
       y = expression(frac("Intron TPM", 
                           "Exon + Intron TPM"))) +
  geom_vline(xintercept = 10.5, linetype = 'dashed') + 
  scale_fill_manual(values = stage_cols) +
  theme(
    # axis.text.x = element_text(angle = 90,
    #                            vjust = 0.5,
    #                            hjust = 0.5),
    legend.position = "none"
  )

save_plot(file.path(fig_dir, "xenopus_t-pre-mRNA-ratios.pdf"),
                    p,
          base_asp = 3)

p

save_plot(file.path(pfig_dir$fig_1, "1b_xenopus_t-pre-mRNA-ratios.pdf"),
           p,
           base_asp = 2)

p
```

```{r}
sessioninfo::session_info()
```
