---
title: "Long-read characterization of the MZT in Drosophila"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, message = FALSE)
source(here::here("R/globals.r"))
library(Rsamtools)
```



RNA was collected from 0-1 hour or 4-5 hour *Drosophila* embryos and sequenced using direct RNA sequencing on the oxford nanopore technologies (ONT) platform.

## Goals:

- Examine isoforms in the maternal transcriptome and post major-wave of the MZT. Compare to isoforms detected by short read sequencing that have evidence of switching during the MZT. 

- Examine poly(A) tail lengths and RNA modifications pre and post major-wave estimated from raw sequencing traces. 

## Isoform characterization:

### Approach 

- Nanopore reads were rebasecalled offline using guppy (necessary for calling polyA tails).
- reads were aligned to either the transcriptome or genome (splice-aligned) using minimap2 with appropriate settings

- transcripts were quantified with salmon using the transcriptome alignments 

```{r}
fig_dir <- file.path("figs_ont")
dir.create(fig_dir, showWarnings = FALSE)
col_pal <- scbp::palette_OkabeIto
```

### Library QC

```{r, fig.cap="Summary of read alignments to the drosophila transcriptome (using minimap2) or to the yeast control RNA. Numbers at the base of each bar indicate total # of raw reads."}

samples <- c(
  "0-1_hr_1",
  "0-1_hr_2",
  "0-1_hr_3",
  "0-1_hr_1_rerun",
  "4-5_hr_1",
  "4-5_hr_2",
  "4-5_hr_3",
  "4-5_hr_1_rerun"
)

bam_fns <- file.path(project_dir, 
                     "data", 
                     "ont",
                     "sorted_alignments",
                     str_c(samples, ".bam"))

# from Rsamtools tutorial
.unlist <- function (x) {
  ## do.call(c, ...) coerces factor to integer, which is undesired
  x1 <- x[[1L]]
  if (is.factor(x1)) {
    structure(unlist(x), class = "factor", levels = levels(x1))
  } else {
    do.call(c, x)
  }
}


bam_align_summary <- function(bam_fn,
                              fields = c("qname", "flag", "rname")){
  params <- ScanBamParam(what = fields)
  bam <- scanBam(bam_fn, 
               param = params)
  bam <- unname(bam) # names not useful in unlisted result
  elts <- setNames(bamWhat(params), bamWhat(params))
  lst <- lapply(elts, function(elt) .unlist(lapply(bam, "[[", elt)))
  b <-  as_tibble(do.call("data.frame", lst)) 
  n_reads <- length(unique(b$qname))
  aligned <- filter(b,flag != 4)
  n_fly_aligned <- filter(aligned, rname != "YHR174W") %>% 
    pull(qname) %>% 
    unique() %>% 
    length()
  n_yeast_aligned <- filter(aligned, rname == "YHR174W") %>% 
    pull(qname) %>% 
    unique() %>% 
    length()
  percent_aligned <- 100 * (n_fly_aligned / n_reads)
  percent_yeast_aligned <- 100 * (n_yeast_aligned / n_reads)
  
  if(is.null(names(bam_fn))){
    sample <- basename(bam_fn) %>% str_remove(".bam")
  } else {
    sample <- names(bam_fn)[1]
  }
  
  tibble(sample, 
         n_reads, 
         n_fly_aligned, 
         percent_aligned, 
         n_yeast_aligned, 
         percent_yeast_aligned,
         bam_fn)
}

fly_alignments <- map_dfr(bam_fns, 
                          bam_align_summary)

p <- fly_alignments %>% 
  mutate(percent_not_aligned = 100 - percent_aligned - percent_yeast_aligned) %>% 
  select(-bam_fn) %>% 
  pivot_longer(cols = -sample) %>% 
  filter(!name %in% c("n_fly_aligned", "n_yeast_aligned", "n_reads")) %>% 
  mutate(name = str_remove(name, "percent_") %>% 
           str_to_title() %>% 
           str_replace("_", " ")) %>% 
  ggplot(aes(sample, value)) +
    geom_col(aes(fill = name)) +
  geom_text(data = fly_alignments %>% select(sample, n_reads),
            aes(x = sample,
                y = 10,
                label = scales::comma(n_reads)),
                angle = 90) +
  scale_fill_manual(values = col_pal) +
  labs(x = "Library",
       y = "Percent of Reads",
       fill = "") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1,
                                   vjust = 0.5))
p
save_plot(file.path(fig_dir, "alignment-summary.pdf"), p,
          base_asp = 1.6,
          base_height = 5.25)
```

### Read lengths

```{r, fig.cap="Read length distributions of reads that map to the drosophila transcriptome."}

read_length_summary <- function(bam_fn){
  b <- scanBam(bam_fn, 
               param = ScanBamParam(what = c("qname", "flag", "rname", "seq"),
                                    flag = scanBamFlag(isUnmappedQuery = FALSE)))
  b[[1]]$rname <- as.character(b[[1]]$rname)
  df <-  as_tibble(do.call(cbind, b[[1]][1:3])) 
  df$read_width <- width(b[[1]]$seq)
  df <- filter(df, rname != "YHR174W")
  df <- group_by(df, qname) %>% summarize(read_width = max(read_width),
                                          .groups = "drop")

   if(is.null(names(bam_fn))){
    df$sample <- basename(bam_fn) %>% str_remove(".bam")
  } else {
    df$sample <- names(bam_fn)[1]
  }
  select(df, sample, qname, read_width)
}
  
read_lengths <- map_dfr(bam_fns, read_length_summary) 

read_lengths_summary <- group_by(read_lengths, sample) %>% 
  summarize(median_vals = median(read_width), .groups = "drop")

p <- ggplot(read_lengths, aes(x = read_width)) +
  geom_histogram(bins = 100) +
  geom_vline(data = read_lengths_summary,
             aes(xintercept = median_vals),
             linetype = "dashed",
             color = 'grey') + 
  facet_grid(sample ~ ., scale = "free_y") +
  labs(x = "Read Length") +
  scale_x_continuous(n.breaks = 10) +
  theme(strip.text = element_text(size= 8))

p

save_plot(file.path(fig_dir, "read_length_distribution.pdf"),
          p,
          base_height = 7,
          base_asp = 1)
```

```{r}
read_lengths_summary
```

## Transcripts detected 

```{r}
salmon_fns <- file.path(project_dir, 
                        "data",
                        "ont",
                        "counts",
                        str_c(samples, "_salmon"),
                        "quant.sf")

tx <- tximport(salmon_fns, 
               type = "salmon",
               txOut = TRUE)

tx <- lapply(tx, function(x){
  if(length(x) == 1){
    return(x)
  } 
  x[rownames(x) != "YHR174W", ]
})
```

```{r}
tx_counts <- tx$counts
colnames(tx_counts) <- samples
tx_summary <- tx_counts %>% 
  as.data.frame() %>% 
  rownames_to_column("tx") %>% 
  pivot_longer(cols = -tx,
               names_to = "sample",
               values_to = "ont_counts") %>% 
  mutate(developmental_stage = str_remove(sample, "_[0-9].*$") %>% # match stage names with short read data
           str_replace("_", " ") %>% 
           str_c(., " embryo")) 
  
```

```{r}
p <- tx_summary %>% 
  filter(ont_counts > 0, ont_counts < 500) %>% 
  ggplot(aes(ont_counts)) + 
  geom_histogram(bins = 100) +
  facet_wrap(~sample) + 
  labs(x = "ONT read counts per transcript")

save_plot(file.path(fig_dir, "ont_transcript_counts.pdf"),
          p,
          nrow = 1,
          ncol = 3, 
          base_asp = 0.6)

p
```

```{r}
mdata <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata <- dplyr::select(mdata,
                        Run_s, 
                        developmental_stage_s,
                        rip_antibody_s,
                        source_name_s,
                        strain_s)

# select relevant libs
mdata <- dplyr::filter(mdata,
                        source_name_s == "Embryo",
                        strain_s %in% c("w1118", "eGFP-ME31B"),
                        rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)


#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

fa_type <- "eisa_masked_bt2"

pdata <- mutate(mdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"),
                strain_id = str_replace(strain, "-", "_"))

rissland_files <- file.path(data_dir,
                   "salmon_bt2_masked", 
                   "drosophila",
                   "RISSLAND",
                   fa_type,
                   pdata$Run,
                   "quant.sf")

tids <- read_tsv(rissland_files[1]) %>% 
  dplyr::select(Name)

gtf_fn <- annots$drosophila$gtf
gtf <- import(gtf_fn)
gtf <- as.data.frame(gtf)

tx2gene <- gtf %>% 
  filter(type == "transcript") %>% 
  dplyr::select(transcript_id, gene_id) %>% 
  as_tibble() %>% 
  unique()

tx2gene_mapping <- full_join(tids, tx2gene,
                             by = c("Name" = "transcript_id"))

gene2symbol <- gtf %>% 
  filter(type == "gene") %>% 
  dplyr::select(gene_id, gene_name) %>% 
  as_tibble() %>% 
  unique()

txi <- tximport(rissland_files,
                type="salmon", 
                txOut=TRUE)

cts <- txi$abundance
cts <- cts[!str_detect(rownames(cts), "-I[0-9]*$"), ]
cts <- cts[rowSums(cts) > 0,]

colnames(cts) <- pdata$Run
il_tx_summary <- as.data.frame(cts) %>% 
  rownames_to_column("tx") %>% 
  pivot_longer(cols = -tx,
               names_to = "Run",
               values_to = "counts") %>% 
  left_join(pdata, by = "Run") %>% 
  group_by(developmental_stage, tx) %>% 
  summarize(tpm = mean(counts), .groups = "drop")
```


```{r, fig.cap="Comparision of transcripts detected by Illumina or ONT sequencing. The TPM values from illumina libaries are shown, with transcripts detected by ONT sequencing (>= 1 read) highlighted."}
plt_dat <- inner_join(il_tx_summary, 
          tx_summary, 
          by = c("developmental_stage", "tx")) %>% 
  filter(tpm > 0) %>% 
  mutate(detected_by_nanopore = ont_counts > 0,
         log_tpm = log10(tpm + 0.001))

ont_summary <- filter(plt_dat, ont_counts > 0) %>% 
  group_by(sample) %>% 
  summarize(n = n(), 
            n = str_c("n = ", scales::comma(n)),
            x = 0.0001,
            y = 500,
            .groups = "drop")

p <- ggplot(plt_dat, aes(tpm)) +
  geom_histogram(aes(fill = detected_by_nanopore),
                 bins = 100) +
  scale_x_log10(labels = function(x) {scales::comma(signif(x, digits = 1))}, 
                n.breaks = 7) +
  geom_text(data = ont_summary,
            aes(group = sample,
                x = x,
                y = y,
                label = n)) +
  scale_fill_manual(values = col_pal) +
  facet_grid(sample ~ ., scales = "free_y") +
  labs(x = "Illumina RNA-Seq TPM",
       y = "# of transcripts",
       fill = "Detected in\nONT library") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust =1,
                                   vjust = 0.5))

save_plot(file.path(fig_dir, "transcript_detection_illumina_comparison.pdf"),
          p,
          nrow = 2,
          ncol = 1, 
          base_asp = 1.5)
p

```


### Calculate # of switching-isoforms detected in ONT data

Table showing the # of transcripts that have altered isoform ratios pre and post-major-wave MZT based on illumina libraries, that are detectable by ONT sequencing. ~80-90% of the transcripts are undetectable at both stages. 
```{r}
switch_isos <- read_excel(file.path("..",
                                    "2020-05-24",
                                    "output_files",
                                    "isoform_summary.xlsx"),
           sheet = 2) 
  
switch_isos_summary <- tx_summary %>% 
  mutate(detected = ont_counts > 0) %>% 
  group_by(developmental_stage, tx) %>% 
  summarize(n_samples = sum(detected),
            mean_counts = mean(ont_counts),
            total_counts = sum(ont_counts),
            .groups = "drop") %>% 
  filter(n_samples >= 2) %>% 
  select(developmental_stage, mean_counts, tx) %>% 
  pivot_wider(names_from = "developmental_stage",
              values_from = "mean_counts") %>% 
  left_join(switch_isos, ., by = c("transcript_id"= "tx"))

mutate(switch_isos_summary, 
       `detected_0-1_hr` = ifelse(is.na(`0-1 hr embryo`),
                                FALSE,
                                `0-1 hr embryo` > 0),
       `detected_4-5_hr` = ifelse(is.na(`4-5 hr embryo`),
                                FALSE,
                                `4-5 hr embryo` > 0)) %>% 
  group_by(transcript_type,
           `detected_0-1_hr`,
           `detected_4-5_hr`) %>% 
  summarize(`# of transcripts` = n()) %>% 
  group_by(transcript_type) %>% 
  mutate(percent = 100 * (`# of transcripts`  / sum(`# of transcripts`))) %>% 
  arrange(desc(`# of transcripts`), 
          .by_group = TRUE)
```

```{r}
library(eulerr)
expr_tx_summary <- tx_summary %>% 
  filter(ont_counts > 0) %>% 
  separate(sample, c("timepoint", "tmp", "rep"), sep = "_") 

detected_tx <- split(expr_tx_summary, 
      expr_tx_summary$rep) %>%
  map(., ~split(.x, 
                .x$timepoint) %>%
        map(~pull(.x, tx) %>% unique())) 

comb_detected_tx <- split(expr_tx_summary, 
                expr_tx_summary$timepoint) %>%
        map(~pull(.x, tx) %>% unique()) 

detected_tx <- c(detected_tx, all = list(comb_detected_tx))
plts <- imap(detected_tx, 
    ~euler(.x) %>% 
      plot(., 
           quantities = TRUE,
           fills = col_pal[1:2],
           main = str_c("replicate ", .y)))


pdf(file.path(fig_dir, "transcript_overlap_timepoints.pdf"),
    height = 9, width = 7)
gridExtra::grid.arrange(plts[[1]], plts[[2]], plts[[3]], plts[[4]],
                        ncol= 1)
dev.off()
```

## Examine poly(A) tails

`tailfindr` 

Generate a table with each each an the respective alignments to match against the tailfindr polyA tail estimates. Primary alignments  will be selected for matching against the transcripts. 

```{r}
fq_libs <- c("JH_w1118_0-1_011220",
             "JH_w1118_0-1_2nd_091220",
             "w1118_0-1_JH260121",
             "JH_w1118_4-5_081220",
             "JH_w1118_4-5_2nd_091220",
             "w1118_4-5_JH260121")

names(fq_libs) <- c(
  "0-1_hr_1",
  "0-1_hr_2",
  "0-1_hr_3",
  "4-5_hr_1",
  "4-5_hr_2",
  "4-5_hr_3"
)

tail_fns <- file.path(project_dir, 
                     "data", 
                     "ont",
                     "offline-basecalling",
                     fq_libs,
                     "workspace",
                     "pAtails",
                     str_c(fq_libs, ".csv"))

# map tail lengths to transcripts
# note that this function will just keep primary alignments 
# no MAPQ filtering by default
map_tails_to_tx <- function(tail_fn,
                            bam_fn, 
                            flags_to_exclude = c(4, 256, 272, 2048),
                            min_mapq = 0){
  b <- scanBam(bam_fn, 
               param = ScanBamParam(what = c("qname", "flag", "rname", "mapq"),
                                    flag = scanBamFlag(isUnmappedQuery = FALSE)))
  # prevent factor coercion to integer
  b[[1]]$rname <- as.character(b[[1]]$rname)
  read_df <-  as_tibble(do.call(cbind, b[[1]][1:4])) 
  tail_df <- read_csv(tail_fn) %>% dplyr::rename(qname = read_id)
  res <- inner_join(read_df, tail_df) %>% filter(rname != "YHR174W")
  res <- mutate(res, 
                sample = basename(tail_fn) %>% str_remove(".csv"),
                .before = 1)
  filter(res, 
         !flag %in% flags_to_exclude,
         mapq >= min_mapq)
}

tail_lens <- map2_dfr(tail_fns, bam_fns, map_tails_to_tx) %>% 
  filter(!is.na(tail_length))

fq_lookup <- names(fq_libs)
names(fq_lookup) <- fq_libs

tx2symbol <- gtf %>% 
  filter(type == "transcript") %>% 
  dplyr::select(transcript_id, gene_name) %>% 
  as_tibble() %>% 
  unique()

tail_lens <- left_join(tail_lens, tx2symbol, by = c("rname" = "transcript_id"))
```



```{r}
tail_lens <- tail_lens %>% 
  mutate(sample_id = fq_lookup[sample]) %>% 
  separate(sample_id, c("timepoint", "tmp", "rep"), sep = "_", remove = FALSE) %>% 
  dplyr::select(-tmp) 

pal_sample_summary <- tail_lens %>% 
  group_by(sample, timepoint, rep, rname) %>%
  summarize(median_tail_len = median(tail_length, na.rm = TRUE),
            n_reads = n(),
            gene_name = unique(gene_name)) %>% 
  ungroup() %>% 
  filter(n_reads > 5) %>% 
  select(-sample, -n_reads) 
   
pal_all_summary <- tail_lens %>% 
  mutate(sample = "all", 
         rep = "all") %>% 
  group_by(sample, timepoint, rep, rname) %>%
  summarize(median_tail_len = median(tail_length, na.rm = TRUE),
             n_reads = n(),
            gene_name = unique(gene_name)) %>% 
  ungroup()  %>% 
  filter(n_reads > 5) %>% 
  select(-sample, -n_reads)

pal_all_summary <- bind_rows(pal_sample_summary, pal_all_summary)
```


```{r}
library(eulerr)
detected_tx <- split(pal_all_summary, 
      pal_all_summary$rep) %>%
  map(., ~split(.x, 
                .x$timepoint) %>%
        map(~pull(.x, rname))) 

plts <- imap(detected_tx, 
    ~euler(.x) %>% 
      plot(., 
           quantities = TRUE,
           fills = col_pal[1:2],
           main = str_c("replicate ", .y)))


pdf(file.path(fig_dir, "pal_overlap_timepoints.pdf"),
    height = 9, width = 7)
gridExtra::grid.arrange(plts[[1]], plts[[2]], plts[[3]], plts[[4]],
                        ncol= 1)
dev.off()
```

```{r}
pal_all_summary_wide <- pal_all_summary %>% 
  pivot_wider(names_from = timepoint, 
              values_from = median_tail_len) 

plt_stats <- pal_all_summary_wide %>% 
  na.omit() %>% 
  mutate(x = 0.8 * max(`0-1`, na.rm = T),
         y = 0.8 * max(`4-5`, na.rm = T)) %>% 
  group_by(rep, x, y) %>% 
  summarize(n = str_c("n = ", scales::comma(n())),
            r = str_c("r = ", signif(cor(`0-1`, `4-5`, method = "spearman", use = "pairwise.complete.obs"), 2)),
            lab = str_c(n, r, sep = "\n"))

p <- ggplot(pal_all_summary_wide, aes(`0-1`, `4-5`)) + 
    geom_point(size = 0.1, alpha = 0.25) + 
    facet_wrap(~rep, nrow = 1) +
    coord_equal() +
    labs(x = "Poly(A) tail length (0-1 hr)",
         y = "Poly(A) tail length (4-5 hr)") +
    geom_text(data = plt_stats, aes(x = x, y = y, label = lab))
    
save_plot(file.path(fig_dir, "polyA_tail_lens_per_rep_5reads.pdf"),
          p,
          nrow = 1,
          ncol = 2, 
          base_asp = 1.5)
p
```


```{r}
plt_dat <- pal_all_summary_wide %>% 
  select(-gene_name) %>% 
  pivot_longer(cols = -c(rname, rep)) 
  
plt_stats <- plt_dat %>% 
  na.omit() %>% 
  group_by(rep, name) %>% 
  summarize(y = median(value, na.rm = T))

p <-  ggplot(plt_dat, aes(name, value)) + 
    geom_violin(aes(fill = name)) +
    scale_fill_manual(values = col_pal) + 
    facet_wrap(~rep, nrow = 1) +
    labs(x = "",
         y = "Poly(A) tail length",
         fill = "") +
  theme(legend.position = "none")

save_plot(file.path(fig_dir, "polyA_tail_lens_violin_5reads.pdf"),
          p,
          nrow = 1,
          ncol = 2, 
          base_asp = 1.5)
p
```


```{r}
pal_df <- pal_all_summary_wide %>% 
  na.omit() %>% 
  filter(rep == "all") %>% 
  mutate(pal_log2fc = log2(`4-5`) - log2(`0-1`)) %>% 
  dplyr::rename(tx_id = rname)

transcript_level_log2_te.tsv.gz <- read_tsv(file.path("..",
                                    "2020-05-24",
                                    "output_files",
                                    "all_tx_analysis.tsv.gz")) %>% 
  filter(count_type == "exon") 

te <- read_tsv("../2019-11-21_rpfs/transcript_level_log2_te.tsv.gz") %>%  
  dplyr::select(-transcript_id) %>% 
  dplyr::rename(tx_id = representative_cds_transcript) %>% 
  distinct()  %>% 
  column_to_rownames("tx_id") %>% 
  .[!rowSums(.) == 0, ] %>% 
  .[!rowSums(is.na(.)) == ncol(.), ] %>% 
  rownames_to_column("tx_id")

combined_df <- Reduce(function(x, y) inner_join(x, y, by = "tx_id"),
                      list(pal_df, expr_data, te))

```

```{r}
p1 <- ggplot(combined_df, aes(`0-1`, `0-1_h`)) + 
  geom_point(size = 0.1) +
  labs(x = "Poly(A) tail length (0-1hr)",
       y = expr(paste("log"[2]," TE (0-1hr)")))

p2 <- ggplot(combined_df, aes(`4-5`, `3-4_h`)) + 
  geom_point(size = 0.1) +
  labs(x = "Poly(A) tail length (4-5hr)",
       y = expr(paste("log"[2]," TE (3-4hr)")))

p3 <- ggplot(combined_df, aes(pal_log2fc, log2_te)) + 
  geom_point(size = 0.1) +
  labs(x = expr(paste("log"[2], " Poly(A) tail length (4-5hr/0-1hr)")),
       y = expr(paste("log"[2]," TE (3-4hr/0-1hr)")))
p <- plot_grid(p1, p2, p3, nrow = 1)

save_plot(file.path(fig_dir, "polyA_tail_v_te_5reads.pdf"),
          p,
          nrow = 1,
          ncol = 3, 
          base_asp = 1.0)
p
```



