---
title: 'build SAF references'
author: 'Kent Riemondy RBI'
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_libs}
library(here)
source(here('R', 'globals.r'))
library(valr)
library(tidyverse)
library(rtracklayer)
library(doParallel)

ref_dir <- here("dbases", "drosophila", "saf")
dir.create(ref_dir, showWarnings = FALSE)
```



# Build Reference for feature count approach

```{r annot}
gtf <- import(annots$drosophila$gtf)
gtf <- as.data.frame(gtf) %>% 
  as_tibble()

fasta_fn <- annots$drosophila$genome

chrom_sizes <- read_tsv(str_c(fasta_fn, 
                              '.fai'), 
                        col_types = 'ci---', 
                        col_names = c('chrom', 'size'))
```

This method is analogous to the method used by the Miller et al 2013. Essentially uniquely aligned reads are counted in either exonic regions or intronic regions. Intronic regions that have read coverage in the earliest pre-MZT stages are masked and excluded from intronic quantification to avoid misannotation from obscuring transcriptional responses. 

```{r get_coverage}
mdata <- read_tsv(
  file.path(
    data_dir,
    "raw_data",
    "drosophila",
    "RISSLAND",
    "GSE98106_run_info_geo.txt"
  )
)

mdata <- dplyr::select(mdata,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata <- dplyr::filter(mdata,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s) %>% 
  filter(developmental_stage_s == "0-1 hr embryo")

bws <- dir(
  file.path(data_dir, "bigwigs", "drosophila", "star", "RISSLAND"),
  full.names = TRUE,
  pattern = ".bw"
)

bws <- str_subset(bws, str_c(mdata$Run_s,
                             collapse = "|"))

fwd_bws <- str_subset(bws, "fwd.bw")
rev_bws <- str_subset(bws, "rev.bw")

fwd_df <- map_dfr(fwd_bws,
                  ~read_bigwig(.x, set_strand = "+")) 

rev_df <- map_dfr(rev_bws,
                  ~read_bigwig(.x, set_strand = "-")) 

# sum coverage
fwd_df <- bed_partition(fwd_df, 
                        score = sum(score))

rev_df <- bed_partition(rev_df, 
                        score = sum(score))

ivl_df <- bind_rows(list(
  "+" = fwd_df,
  "-" = rev_df),
  .id = "strand")

ivl_df <- dplyr::select(ivl_df, 
                        chrom:score,
                        strand)

# Remove exons from regions to mask 
gtf_df <- mutate(gtf,
                 start = start - 1,
                 chrom = as.character(seqnames),
                 strand = as.character(strand)) %>% 
  bed_sort()

exons <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                transcript_id,
                exon_number,
                strand) %>% 
  bed_sort() %>% 
  group_by(strand)

ivl_df <- group_by(ivl_df, strand)
intron_coverage <- bed_subtract(ivl_df, exons)

masked_ivls <- filter(intron_coverage, score > 0) %>% 
  group_by(strand) %>% 
  bed_merge() %>% 
  ungroup()


masked_out <- mutate(masked_ivls,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(masked_out, file.path(ref_dir, "fly_intron_mask.bed"), col_names = F)
```

```{r}
exons <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                transcript_id,
                exon_number,
                strand) %>% 
  mutate(exon_number = as.integer(exon_number)) %>% 
  group_by(chrom, transcript_id, strand) %>% 
  arrange(chrom, start, end, .by_group = T) %>% 
  mutate(id = row_number(),
         max_exon_count = max(exon_number)) %>% 
  bed_sort() %>% 
  ungroup()

simple_exons <- dplyr::select(exons,
                              chrom:end, strand) %>% 
  unique() %>%
  group_by(strand)

intron_coverage <- bed_subtract(ivl_df, simple_exons) %>% 
   filter(score > 0) %>% 
  dplyr::select(-score)

multi_exon_mask <- exons %>%
  filter(max_exon_count > 1, 
         end - start > 200) %>% 
  group_by(chrom, transcript_id, strand) %>% 
  mutate(start = ifelse(id == 1, # first exon, don't trim start
                        start,
                        start + 40),
         end = ifelse(id == max_exon_count, # last exon don't trim end
                      end,
                      end - 40)) %>% 
  ungroup() %>% 
  filter(start < end) %>% # remove exons that are fully masked
  dplyr::select(chrom, 
                start,
                end,
                strand)
  
not_modified_exons <- exons %>% 
  filter(end - start <= 200 | ((end - start) > 200 & max_exon_count == 1)) %>% 
  dplyr::select(chrom, 
                start,
                end,
                strand)

exon_mask <- bind_rows(list(multi_exon_mask,
                            not_modified_exons)) %>% 
  bed_sort() %>% 
  unique()
  
intron_coverage <- bind_rows(intron_coverage, 
                             exon_mask)

masked_ivls <- group_by(intron_coverage, strand) %>% 
  bed_merge() %>% 
  ungroup()

masked_out <- mutate(masked_ivls,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(masked_out,
          file.path(ref_dir, "fly_exon_intron_plus_exons_juncts_mask.bed"),
          col_names = F)
```

```{r}
introns <- filter(gtf_df, type == "transcript") %>% 
  dplyr::select(chrom, start, end, gene_name, strand) %>% 
  unique() %>% 
  group_by(strand) %>% 
  bed_subtract(group_by(simple_exons, strand)) %>% 
  group_by(strand) %>% 
  bed_subtract(group_by(masked_ivls, strand)) %>% 
  unique()

introns_out <- introns %>% 
  mutate(score = 0L) %>% 
  dplyr::select(chrom, start, end, gene_name, score, strand) %>% 
  mutate_if(is.numeric, format, scientific = F) %>% 
  mutate_all(str_trim)

write_tsv(introns_out, file.path(ref_dir, "fly_introns.bed"), col_names = F)
introns_out %>% 
  dplyr::select(GeneID = gene_name, Chr = chrom, Start = start, End = end, Strand = strand) %>% 
  write_tsv(., file.path(ref_dir, "fly_introns.saf"))

exons_out <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                gene_name,
                exon_number,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(exons_out, file.path(ref_dir, "fly_exons.bed"), col_names = F)

exons_out %>% 
  dplyr::select(GeneID = gene_name, Chr = chrom, Start = start, End = end, Strand = strand) %>% 
  write_tsv(., file.path(ref_dir, "fly_exons.saf"))

```

```{r}
sessioninfo::session_info()
```

