---
title: "Zebrafish MZT analysis"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message = FALSE, warning = FALSE, include = FALSE}
source(here::here("R/globals.r"))
fig_dir <- "figs"
species <- "zebrafish"
dir.create(fig_dir, showWarnings = FALSE)
```


## Approach

A paper describing a very high resolution time course of zebrafish was published in eLife [A high-resolution mRNA expression time course of embryonic development in zebrafish](https://elifesciences.org/articles/30860). This dataset will likely provide additional proof of principle data for the intronic analysis performed with drosophila datasets. 

RNA-Seq data (paired-end) from 8 stages surrounding the MZT (or ZGA in zebrafish literature) was downloaded with 5 replicates per time point. The data was processed through salmon similarly to the drosophila data. In contrast to Olivia's drosphila data, these data are poly(A) purified mRNA-Seq libraries, so the intronic signal may be more subtle.  

 
## get metadata

```{r get_metadata}
mdata <- read_tsv(file.path(data_dir, "raw_data", species, 
                            "WHITE", "elife-30860-supp1-v1.tsv")) %>% 
  filter(sequencing == "RNASeq",
         str_detect(stage, "Zygote|Cleavage|Gastrula|Blastula"))
mdata
```

```{r }
gtf <- tidy_gtf(annots[[species]]$gtf)

gene2tx <- gtf %>%
   filter(type == "exon") %>%
   select(gene_id, transcript_id) %>%
   na.omit() %>%
   unique()

gene2intron <- read_tsv(here("dbases-rerun", species, "eisa", "tx2gene.tsv"))
```


## Examine pre-mRNA ratios

```{r read_in_salmon}
fa_types <- c(
  "eisa_masked"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon_bt2_masked", 
                        species, 
                        "WHITE",
                        .x,
                        mdata$public_name,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$public_name
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "public_name"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("public_name"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "-I[0-9]*$"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

dat <- mutate(dat,
              tmp_id = ifelse(count_type == "intron",
                               str_remove(Name, "-I[0-9]*$"),
                               Name))
# add in geneid attribute
tpm_dat <- left_join(dat, gene2tx, 
                 by = c("tmp_id" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

```

```{r}
# sum primary and mature TPMs for each gene 
summary_dat <- dplyr::select(tpm_dat, TPM, gene_id, 
              public_name, count_type, sampleName, stageName) %>% 
  group_by(gene_id, public_name, sampleName,
           stageName, count_type) %>% 
  dplyr::summarize(total_tpm = sum(TPM, na.rm = F)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- summary_dat %>% 
  spread(count_type, total_tpm) %>% 
  mutate(primary_ratio  = intron / (intron + exon))

## calc mean primary ratio
tpm_means <- tpm_summary %>% 
  group_by(gene_id, stageName) %>% 
  summarize(mean_primary = mean(primary_ratio, na.rm = T)) 

tpm_means <- mutate(tpm_means, 
                       stage = factor(stageName, 
                                         levels = c(
                                           "1-cell",
                                           "2-cell",
                                           "128-cell",
                                           "1k-cell",
                                           "Dome",
                                           "50pc-epiboly",
                                           "Shield",
                                           "75pc-epiboly")
                                         )
                       )

tpm_means <- dplyr::ungroup(tpm_means)

p <- ggplot(tpm_means, 
       aes(stage, mean_primary)) +
  geom_boxplot(aes(fill = stage), outlier.shape = NA, coef = 1e10) + 
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_brewer(palette = "Paired") +
  geom_vline(xintercept = 4.5, linetype = 'dashed') + 
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5),
    legend.position = "none"
  )

save_plot(file.path(fig_dir, "zebrafish-pre-mRNA-ratios.pdf"),
                    p,
          base_width = 6,
          base_height = 4.5)

p

save_plot( "../figures/pdfs/fig1/zebrafish-pre-mRNA-ratios.pdf",
           p,
          base_width = 6,
          base_height = 4.5)

p

```


```{r}
sessioninfo::session_info()
```
